framework/src/main/java/org/checkerframework/common/purity/PurityVisitor.java

```diff
@@ -0,0 +1,196 @@
+package org.checkerframework.common.purity;
+
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import java.util.*;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.dataflow.qual.Pure;
+import org.checkerframework.dataflow.util.PurityChecker;
+import org.checkerframework.dataflow.util.PurityUtils;
+import org.checkerframework.framework.source.Result;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.javacutil.Pair;
+import org.checkerframework.javacutil.TreeUtils;
+
+public class PurityVisitor extends BaseTypeVisitor<PurityAnnotatedTypeFactory> {
+
+    public PurityVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public Void visitMethod(MethodTree node, Void p) {
+        boolean anyPurityAnnotation = PurityUtils.hasPurityAnnotation(atypeFactory, node);
+        boolean checkPurityAlways = checker.hasOption("suggestPureMethods");
+        boolean checkPurityAnnotations = checker.hasOption("checkPurityAnnotations");
+
+        if (checkPurityAnnotations && (anyPurityAnnotation || checkPurityAlways)) {
```

Comment:
Similar to the above, we can remove the first checker here, and change this guard to just `anyPurityAnnotation || checkPurityAlways`